qgsproject: add support for base64 encoded url

benoitdm-oslandia opened the pull request:
In QgsDataSourceUri, if the 'url' field references a QGIS server then the 'MAP' parameter can contain an url to the project file. When the project is saved in a postgresql db, the connection url will also contains '&' and '=' characters. One solution is to urlencod twice this url and another solution is to use the encode the project file url to base64 and use the 'base64://' scheme.

closes #31192

benoitdm-oslandia committed:
qgsproject: add support for base64 encoded url

In QgsDataSourceUri, if the 'url' field references a QGIS server then the 'MAP' parameter can contain an url to the project file.
When the project is saved in a postgresql db, the connection url will also contains '&' and '=' characters.
One solution is to use the encode the original project file url with base64 and use the 'base64://' scheme.

See issue #31192

benoitdm-oslandia committed:
fixup! qgsproject: add support for base64 encoded url

nirvn commented:
@benoitdm-oslandia , hey, looked at this earlier today. I feel like if we want a base64 encoding for the MAP= request parameter, it might be better to decode it within the relevant src/server classes instead of handling it within the QgsProject class. Namely, we could simply check for and decode base64 in those two locations:

- https://github.com/qgis/QGIS/blob/master/src/server/qgis_mapserver.cpp#L595
- https://github.com/qgis/QGIS/blob/master/src/server/qgsconfigcache.cpp#L139

Also, what should really just happen here is for the MAP query value to be URL encoded. It's an alternative you're pointing out in the description, and it has the advantage of avoiding obfuscating the MAP source project a bit. 

That said, if you see advantages to base64 that I do not see ATM, I'd suggest we go for the src/server approach I suggested above.

nirvn commented:
@benoitdm-oslandia , I'm not a QGIS server user myself, sounds like you're trying to address something that's worth fixing. I'd still feel better if the base64 encoding/decoding would be handled within src/server itself. Would that be OK with you?

benoitdm-oslandia commented:
@nirvn so you would prefer I apply the decoding outside the `QgsProject::read` function at https://github.com/qgis/QGIS/blob/master/src/server/qgis_mapserver.cpp#L595 and https://github.com/qgis/QGIS/blob/master/src/server/qgsconfigcache.cpp#L139?

benoitdm-oslandia committed:
fixup! qgsproject: add support for base64 encoded url

benoitdm-oslandia commented:
Fixed! But there are still many places where the QgsProject::read function is called like qgslandingpageutils.cpp or qgsprocess.cpp. Not applying the fix at these places fell like "I fixed my issue but I did not look at the possible side effects".

nirvn commented:
@benoitdm-oslandia , ah oui, you should add it in qgslandingpageutils.cpp too -- basically, what I'm thinking here is that it should be a server-specific handling of base64, where the URL context QGIS server is accessed through seems to justify base64 encoding.

But e.g. for QGIS desktop's CLI, or qgis_process CLI, I'm not sure how I'd feel about base64 there :)

Happy to be challenged by others on this though.

cioddi commented:
This appears to be related to the problem I had with WMS sources that contained url-parameter values that include `&` or `=`. The problem is that all values added to QgsDataSourceUri in a Urlencoded way becomes urldecoded the next time it is used (e.g. in a http-request).  

I believe this must be solved at the QgsDataSourceUri level for the solution to cover all supported source types. That way we can be sure a value added to QgsDataSourceUrl isn't altered on retrieval, which currently is the case.


I had a PR open for this that solves the problem, but it became stale and was closed.

https://github.com/qgis/QGIS/pull/59144
